### PR TITLE
Major update: Knex query builder for PG with standard seneca store tests

### PR DESCRIPTION
**Pull request by Jose-Perigolo**
- Added Knex SDK, Postgresql too
- Updated README for Knex use
- Incomplete automated tests
- Created qbuilder.js for cleaner code - communication with Knex
- Updated seneca entity for Knex

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,9 +1,9 @@
 {
-  "name": "seneca-mem-store",
+  "name": "seneca-knex-store",
```
Comment (rjrodger) [inline package.json]: let's use the new naming convention: `@seneca/knex-store`

File: package.json [hunk 2]
```diff
@@ -62,5 +62,9 @@
     "LICENSE",
     "src",
     "dist"
-  ]
+  ],
+  "dependencies": {
```
Comment (rjrodger) [inline package.json]: these should be peer dependencies with '>=', and then also have devDeps for specific versions - we can't control the database sdk version of client code

File: src/knex-store.ts [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+  // Take a reference to the calling Seneca instance
+  let seneca: any = this
+  
+  let conf = intern.getConfig(seneca)
+  knex.conection = conf
+
+  let init = seneca.export('entity/init')
+
+  // merge default options with any provided by the caller
+  options = seneca.util.deepextend(
```
Comment (rjrodger) [inline src/knex-store.ts]: are these options relevant to knex?

File: src/knex-store.ts [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+  // The calling Seneca instance will provide
+  // a description for us on init(), it will
+  // be used in the logs
+  let desc: any
+
+  // Our super awesome in knex database. Please bear in mind
+  // that this store is meant for fast prototyping, using
+  // it for production is not advised!
+  let entmap: any = {}
```
Comment (rjrodger) [inline src/knex-store.ts]: why would the knex need this? the entmap is an internal variable of mem-store - it is not relevant for knex

File: src/intern.ts [hunk 5]
```diff
@@ -1,219 +1,179 @@
+const { Q, transactionQuery} = require('./qbuilder')
```
Comment (rjrodger) [inline src/intern.ts]: there is no need to create transactions in the store - the store operations are single database operations

File: src/knex-store.ts [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+        // create a new entity
+        return intern
+          .insert_knex_transaction(ent, id)
+          .then((result: any) => {
+            // update the entity with the new id
+            ent.id = id
+
+            // save the entity in the entmap
+            entmap[ent.id] = ent
```
Comment (rjrodger) [inline src/knex-store.ts]: no need for this

File: src/knex-store.ts [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+  // Init the store using the seneca instance, merged
+  // options and the store description object above.
+  let meta = init(seneca, options, store)
+  //let meta = seneca.store.init(seneca, options, store)
+
+  // int() returns some metadata for us, one of these is the
+  // description, we'll take a copy of that here.
+  desc = meta.desc
+
+  seneca.add(
```
Comment (rjrodger) [inline src/knex-store.ts]: these are mem-store operations and can be removed

File: test/knex.test.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
+/*
```
Comment (rjrodger) [inline test/knex.test.js]: create a file called `quick.js` in the test folder to quickly validate if things are working

perhaps:
```
let seneca = Seneca()
  .test()
  .use('promisfy')
  .use('entity')
  .use('knex-store', {
    // whatever options you need for knex to talk to the database
  })

await seneca.ready()

// should create a row in table `foo` with columns 'id' and `x` 
let foo1 = await seneca.entity('foo').data$({x:1}).save$()

console.log(foo1) // prints {entity$:'-/-/foo', id:'...', x: 1}

let rows = seneca.entity(foo).list$()
console.log(rows) /// lists rows
```

File: src/knex-store.ts [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+  }
+  generate_id: any
+}
+
+function knex_store(this: any, options: Options) {
+  
+  // Take a reference to the calling Seneca instance
+  let seneca: any = this
+  
+  let conf = intern.getConfig(seneca)
```
Comment (rjrodger) [inline src/knex-store.ts]: doesn't getConfig take a spec as first argument?

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Jose-Perigolo) [reply package.json]: Got it

File: package.json (hunk 2 above)
Comment (Jose-Perigolo) [reply package.json]: Agreed, ok

File: src/knex-store.ts (hunk 4 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: I kept a few lines of mem-store code that I didn't quite understand what it was for, removing

File: src/intern.ts (hunk 5 above)
Comment (Jose-Perigolo) [reply src/intern.ts]: Understood, removing. There is a huge variety of functions present in the Knex QueryBuilder (select, upsert, transacting, returning...), should I add or try to make compatibility of all of them in our store?
[https://knexjs.org/guide/query-builder.html#common](url)

File: src/knex-store.ts (hunk 7 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: I kept a few lines of mem-store code that I didn't quite understand what it was for, removing

File: src/knex-store.ts (hunk 6 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: Ok

File: test/knex.test.js (hunk 8 above)
Comment (Jose-Perigolo) [reply test/knex.test.js]: Yes, I was doing that, I just wasn't sure how to use seneca-test, I'll create the test

File: src/intern.ts (hunk 5 above)
Comment (rjrodger) [reply src/intern.ts]: seneca-entity only has four operations: save, load, list, remove

File: src/knex-store.ts (hunk 4 above)
Comment (rjrodger) [reply src/knex-store.ts]: mem-store is just an in-memory simulation of a database using JS objects - not needed when we have a real database - review the code of https://github.com/senecajs/seneca-postgres-store for example

File: src/intern.ts (hunk 5 above)
Comment (Jose-Perigolo) [reply src/intern.ts]: True, ok

File: src/knex-store.ts (hunk 4 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: This I understood, I said in relation to this line of code, I didn't know that it belonged to the mem-store because I didn't know the purpose of `entmap` variable

File: package.json [hunk 10]
```diff
@@ -19,31 +19,35 @@
     "Efim Bogdanovsky (https://github.com/lilsweetcaligula)"
```
Comment (rjrodger) [inline package.json]: replace with yourself! :)

File: package.json [hunk 11]
```diff
@@ -19,31 +19,35 @@
...
-    "memory"
+    "knex",
+    "knexjs"
   ],
   "peerDependencies": {
     "seneca": ">=3",
-    "seneca-entity": ">=18"
+    "seneca-entity": ">=18",
+    "knex": ">=2.4.2",
+    "pg": ">=8.8.0"
```
Comment (rjrodger) [inline package.json]: pg should be a devDep - knex works with other dbs too

File: src/intern.ts [hunk 12]
```diff
@@ -1,219 +1,151 @@
...
 
-  static is_object(x: any): boolean {
-    return '[object Object]' === toString.call(x)
+    return conf
   }
 
-  static is_date(x: any): boolean {
-    return '[object Date]' === toString.call(x)
-  }
+  static buildCtx(seneca: any, msg: any, meta: any) {
```
Comment (rjrodger) [inline src/intern.ts:177]: remove in this version - we will add tx support separately

File: src/knex-store.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+      return is_new ? do_create() : do_save()
+
+      // Create a new entity
+      async function do_create() {
+        // generate a new id for the entity
+        const id = options.generate_id(ent)
+
+        // create a new entity
+        try {
+          const result = await intern
```
Comment (rjrodger) [inline src/knex-store.ts]: what is the `result` here?

File: test/config/database/config.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+function getConfig() {
+  if (process.env.CI) {
```
Comment (rjrodger) [inline test/config/database/config.ts:2]: use a namespace safe env var: `SENECA_KNEX_TEST_CONFIG`

File: test/knex.test.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+// const Shared = require('seneca-store-test')
+const Lab = require('@hapi/lab')
+const lab = (exports.lab = Lab.script())
+const { describe, it, before } = lab
+const { expect } = require('@hapi/code')
+
+const KnexStore = require('../src/knex-store')
+const DbConfig = require('./config/database/config')
+
+describe('knex-store tests', function () {
```
Comment (rjrodger) [inline test/knex.test.js:58]: we will also need to run the standard store tests
see: https://github.com/senecajs/seneca-postgres-store/blob/master/test/postgres.test.js#L28

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: test/quick.ts [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    const foo1 = await s0.entity('foo').data$({ x: 5 }).remove$()
+    console.log(foo1)
+  }
+
+  awaitSeneca().then(() => {
+    quickSave().then(() => {
+      quickList().then(() => {
+        quickUpdate().then(() => {
+          quickRemove().then(() => {
+            console.log('done')
```
Comment (lilsweetcaligula) [inline test/quick.ts]: Not a request, but in the future, - if you have a pattern like this, - you could inline it using an IIFE (that's a buzzword for anonymous function), like so:
```javascript
; (async () => {
  await awaitSeneca()
  await quickSave()
  await quickList()
  await quickUpdate()
  await quickRemove()
  console.log('done')
})()
```

Comment (Jose-Perigolo) [reply test/quick.ts]: Right, in fact much cleaner, I'll do it here

File: src/knex-store.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    // use in seneca.use(), eg seneca.use('knex-store').
+    name: STORE_NAME,
+
+    save: async function (this: any, msg: any, meta: any, reply: any) {
+      // Take a reference to Seneca
+      // and the entity to save
+      const seneca = this
+      const { ent, q } = msg
+      console.log(ent)
+      console.log(meta)
```
Comment (lilsweetcaligula) [inline src/knex-store.ts]: In client code using this plugin, these `console.log`'s will produce unwanted logs. Is there a reason you included them in this PR?

File: src/knex-store.ts [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+      // Create a new entity
+      async function do_create() {
+        // generate a new id for the entity
+        const id = options.generate_id(ent)
+
+        // create a new entity
+        try {
+          const result = await intern
+            .insertKnex(ent, id)
+            console.log(result)
```
Comment (lilsweetcaligula) [inline src/knex-store.ts]: In client code using this plugin, this `console.log` will produce unwanted logs. Is there a reason you included it in this PR?

File: src/knex-store.ts [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+        } catch (err) {
+          // if there is an error, call the reply
+          // callback with the error
+          reply(err)
+        }
+      }
+
+      // Save an existing entity  
+      function do_save() {
+        intern.updateKnex(ent, ctx, q)
```
Comment (lilsweetcaligula) [inline src/knex-store.ts]: You may have forgotten to await on the promise here? E.g.:
```javascript
await intern.updateKnex(ent, ctx, q)
```

File: src/knex-store.ts [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+
+      // Create a new entity
+      async function do_create() {
+        // generate a new id for the entity
+        const id = options.generate_id(ent)
+
+        // create a new entity
+        try {
+          const result = await intern
+            .insertKnex(ent, id)
```
Comment (lilsweetcaligula) [inline src/knex-store.ts]: This doesn't seem to save entity data to the database.

File: test/knex.test.js [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+  before(() => {
+    return new Promise((done) => {
+      seneca().ready(done)
+    })
+  })
+
+  it('save', async () => {
+    const s0 = await seneca().entity().begin()
+    console.log(s0)
+    const foo1 = await s0.entity('foo').data$({ x: 1 }).save$()
```
Comment (lilsweetcaligula) [inline test/knex.test.js]: In this test, in addition to the existing code, - please make an explicit call to the database using `knex` directly, and verify that the `foo` entity has been saved with the correct data, e.g.:
```javascript
const foo_record = await knex('foo').first()
expect(foo_record).to.be.an.object()
expect(foo_record.x).to.equal(1)
```

File: test/knex.test.js [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+
+    expect('foo').to.exist()
+    expect(typeof 'foo'.id).to.equal('number')
+    expect('foo'.x).to.equal(5)
+  })
+
+  it('remove', async () => {
+    const s0 = await seneca().entity.begin()
+    console.log(s0)
+    const foo1 = await s0.entity('foo').data$({ x: 5 }).remove$()
```
Comment (lilsweetcaligula) [inline test/knex.test.js]: In this test, in addition to the existing code, - please make two explicit calls to the database using `knex` directly - one before the call to `.remove$`, and one after the call.

The first call should get the count of how many `foo` records are in the database at that time.
The second call should `expect` that after `.remove$` was called - the count of `foo` records in the database is one less.

File: test/knex.test.js [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+const DbConfig = require('./config/database/config')
+
+describe('knex-store tests', function () {
+  const test = seneca()
+    .test()
+    .use('entity')
+    .use(KnexStore, { DbConfig })
+    .use('promisify')
+
+  console.log(test)
```
Comment (lilsweetcaligula) [inline test/knex.test.js]: Please do not leave `console.log`'s in automated tests.

File: test/knex.test.js [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+
+  before(() => {
+    return new Promise((done) => {
+      seneca().ready(done)
+    })
+  })
+
+  it('save', async () => {
+    const s0 = await seneca().entity().begin()
+    console.log(s0)
```
Comment (lilsweetcaligula) [inline test/knex.test.js]: Please do not leave `console.log`'s in automated tests.

File: test/knex.test.js [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    expect('foo'.x).to.equal(5)
+  })
+
+  it('remove', async () => {
+    const s0 = await seneca().entity.begin()
+    console.log(s0)
+    const foo1 = await s0.entity('foo').data$({ x: 5 }).remove$()
+    console.log(foo1)
+
+    expect('foo').to.exist()
```
Comment (lilsweetcaligula) [inline test/knex.test.js]: Erm, what exactly are you trying to achieve with this line of code?
P. S. For reference: https://hapi.dev/module/code/api/?v=9.0.3#exist

File: test/knex.test.js [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+    expect(typeof 'foo'.id).to.equal('number')
+    expect('foo'.x).to.be.false
```
Comment (lilsweetcaligula) [inline test/knex.test.js]: Why would `'foo'.x` be `false`?
P. S. Besides, I believe you may have forgotten to call `.false()`. For reference: https://hapi.dev/module/code/api/?v=9.0.3#false

File: src/intern.ts [hunk 27]
```diff
@@ -1,219 +1,151 @@
...
-          }
-        }
+    return false
+  }
+
+  static getConfig(spec: any) {
+    let conf
+
+    if ('string' === typeof spec) {
+      const urlM = /^postgres:\/\/((.*?):(.*?)@)?(.*?)(:?(\d+))?\/(.*?)$/.exec(spec)
```
Comment (lilsweetcaligula) [inline src/intern.ts:145]: Please see if you could use `url.parse` here instead. E.g.
https://github.com/brianc/node-postgres/tree/master/packages/pg-pool#create

File: test/quick.ts [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+
+  console.log(test)
+
+  async function awaitSeneca() {
+      await seneca().ready()
+  }
+
+   async function quickSave() {
+    const s0 = await seneca().entity().begin()
+    console.log(s0)
```
Comment (lilsweetcaligula) [inline test/quick.ts:22]: Please do not leave `console.log`'s in automated tests.

File: test/quick.ts [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+const DbConfig = require('./config/database/config')
+
+
+  const test = seneca()
+    .test()
+    .use('entity')
+    .use(KnexStore, { DbConfig })
+    .use('promisify')
+
+  console.log(test)
```
Comment (lilsweetcaligula) [inline test/quick.ts:14]: Please do not leave `console.log`'s in automated tests.

**Review by lilsweetcaligula — CHANGES_REQUESTED** [review]

File: src/knex-store.ts (hunk 17 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: It was just for testing, no reason, I'll take it off

File: src/knex-store.ts (hunk 18 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: No reason, deleted.

File: src/knex-store.ts (hunk 19 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: You're right! Its a promise

File: src/knex-store.ts (hunk 20 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts]: Right, missing one argument

File: test/knex.test.js (hunk 21 above)
Comment (Jose-Perigolo) [reply test/knex.test.js]: Ok

File: test/quick.ts (hunk 16 above)
Comment (rjrodger) [reply test/quick.ts]: yes - in general prefer async await over the Promise api

File: package.json (hunk 10 above)
Comment (Jose-Perigolo) [reply package.json]: Ok

File: src/intern.ts (hunk 27 above)
Comment (rjrodger) [reply src/intern.ts:145]: actually since, knex works with multiple databases, there should be no need for this at all
@Jose-Perigolo you should have a `dbconf` option that can be used to provide database specific config - we don't care what is inside it

File: package.json (hunk 11 above)
Comment (Jose-Perigolo) [reply package.json]: True, in the future we can use knex-store to adapt to other DBs?

File: src/intern.ts (hunk 12 above)
Comment (Jose-Perigolo) [reply src/intern.ts:177]: Ok

File: test/config/database/config.ts (hunk 14 above)
Comment (Jose-Perigolo) [reply test/config/database/config.ts:2]: Got it

File: test/knex.test.js (hunk 15 above)
Comment (Jose-Perigolo) [reply test/knex.test.js:11]: That was done, I will commit the change

Comment (Jose-Perigolo) [reply test/knex.test.js:58]: Ok

File: src/knex-store.ts [hunk 30]
```diff
@@ -0,0 +1,153 @@
+/* Copyright (c) 2010-2022 Richard Rodger and other contributors, MIT License */
+import { intern } from './intern'
+const Pg = require('pg')
```
Comment (rjrodger) [inline src/knex-store.ts:3]: I kinda missed this - it's a big one :)
knex-store needs to work with any database that knex can work with
that means you can't reference pg anywhere except in test code

File: src/intern.ts (hunk 27 above)
Comment (Jose-Perigolo) [reply src/intern.ts:145]: I thought we were going to use knex-store just for postgre, I can do that yes, it will be better

File: src/knex-store.ts (hunk 30 above)
Comment (Jose-Perigolo) [reply src/knex-store.ts:3]: Ok!

File: src/qbuilder.ts [hunk 31]
```diff
@@ -0,0 +1,52 @@
+
+const knex = require('knex')({
```
Comment (rjrodger) [inline src/qbuilder.ts:1]: this configuration should be passed in using a `bdconf` option to the plugin, to hide the actual database

Comment (rjrodger) [reply src/qbuilder.ts:1]: also in this code, you are create a universal single instance of knex, which is not safe - the knex instance should be inside the plugin definition function - see https://github.com/senecajs/seneca-postgres-store/blob/master/postgresql-store.js#L21

File: test/quick.ts (hunk 29 above)
Comment (rjrodger) [reply test/quick.ts:14]: it's actually ok here - quick.js is meant to be a "quick" script to smoke check the lib - it's not part of the test suite
